### PR TITLE
Check "var" keyword when varType is null or JCErroneous

### PR DESCRIPTION
**Pull request by seonwoo960000**
## What's changed?
When `var` is used with unknown type, rewrite consider `var` as `Space`. We need to add additional logic to identifyt `var` keyword. 

## What's your motivation?
[Issue link](https://github.com/openrewrite/rewrite/issues/3744)

Below test code doesn't work. It rewrites `var a = b` into `finalvar a = b` (should have space between final and var) 
```java
@Test
@Issue("https://github.com/openrewrite/rewrite/issues/3744")
void missingTypePlusVarNotMissingSpace() {
    rewriteRun(
      s -> s.typeValidationOptions(TypeValidation.none()),
      //language=java
      java(
        """
          import foo.Unknown;
          class Bar {
              void test(Unknown b) {
                  var a = b;
              }
          }
            """,
        """
          import foo.Unknown;
          class Bar {
              void test(Unknown b) {
                  final var a = b;
              } 
          }
            """
      )
    );
}
```

## Anything in particular you'd like reviewers to focus on?
- Whether it's ok to put the checking logic into
  - `ReloadableJava11ParserVisitor`
  - `ReloadableJava17ParserVisitor`
  - `ReloadableJava21ParserVisitor`

## Anyone you would like to review specifically?
N/A 

## Have you considered any alternatives or workarounds?
N/A 

## Any additional context
AS-IS 
image

TO-BE 
image

### Checklist
- [X] I've added unit tests to cover both positive and negative cases
- [X] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [X] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java [hunk 1]
```diff
@@ -1507,9 +1507,19 @@ private J.VariableDeclarations visitVariables(List<VariableTree> nodes, Space fm
...
+        } else if (vartype instanceof JCErroneous) {
+            try {
+                if (source.substring(node.startPos, node.pos).startsWith("var")) {
+                    // "var x = unknownType" can be translated into "(ERROR) var x = unknownType"
+                    typeExpr = new J.Identifier(randomId(), sourceBefore("var"), Markers.EMPTY, emptyList(), "var", typeMapping.type(vartype), null);
+                    typeExpr = typeExpr.withMarkers(typeExpr.getMarkers().add(JavaVarKeyword.build()));
+                }
+            } catch (Exception e) {
+                typeExpr = null;
+            }
```
Comment (timtebeek) [inline rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java:1522]: @knutwannheden is probably the better judge of any possible adverse effects if we were to adopt this, but I found it also works without the extra conditional block, by falling into the var-handling else block just below.

Should we keep this try-catch or can we rely on the existing logic, effectively only removing the ` || vartype instanceof JCErroneous` above?
```suggestion
```

**Review by timtebeek — COMMENTED** [review]
Thanks a lot @seonwoo960000 ! Promising work; I've tagged @knutwannheden for an additional look following my suggestion, but this this looks good overall already.

File: rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java [hunk 2]
```diff
@@ -1555,6 +1565,7 @@ private J.VariableDeclarations visitVariables(List<VariableTree> nodes, Space fm
         for (int i = 0; i < nodes.size(); i++) {
             JCVariableDecl n = (JCVariableDecl) nodes.get(i);
 
+            // how to capture var ?
```
[the remark this answers is not in the document]
Comment (seonwoo960000) [reply rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java:1568]: Ah, missed that point. I'll remove it right away.

File: rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java (hunk 1 above)
Comment (seonwoo960000) [reply rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java:1522]: Looks cleaner. Without any adverse effects, let's apply your suggestion.

**Comment by knutwannheden** [comment]
Can we have a test for this in rewrite-java-tck?

**Comment by traceyyoshima** [comment]
@knutwannheden @timtebeek I'm going to take a quick look at the node from the 11, 17, and 21 parser. The compiler should know it's a `var` regardless of an erroneous type. So, I suspect it's possible to check the symbol on the node rather than compare the source position.

**Comment by traceyyoshima** [comment]
> @knutwannheden @timtebeek I'm going to take a quick look at the node from the 11, 17, and 21 parser. The compiler should know it's a `var` regardless of an erroneous type. So, I suspect it's possible to check the symbol on the node rather than compare the source position.

We'll need to use the source code at the cursor position to detect `var`.

@seonwoo960000, please add a test to [VariableDeclarationsTest](https://github.com/openrewrite/rewrite/blob/main/rewrite-java-tck/src/main/java/org/openrewrite/java/tree/VariableDeclarationsTest.java#L64C22-L64C22)

```java
    @Test
    @MinimumJava11
    void unknownVar() {
        rewriteRun(
          spec -> spec.typeValidationOptions(TypeValidation.none()),
          java(
            """
              class Test {
                  void test(Unknown b) {
                      final var a = b;
                  }
              }
              """
          )
        );
    }
```

**Comment by seonwoo960000** [comment]
> Can we have a test for this in rewrite-java-tck?

Hi, @knutwannheden 😄 . Should I add the same tests in `rewrite-java-tck` as well?

**Comment by seonwoo960000** [comment]
Hi @traceyyoshima , I've added the test you provided!

**Comment by traceyyoshima** [comment]
> Hi @traceyyoshima , I've added the test you provided!

Thank you!

> Hi, @knutwannheden 😄 . Should I add the same tests in rewrite-java-tck as well?

Ah, the `VariableDeclarationTest` is located in `rewrite-java-tck`. I agreed with Knut and saved you a few steps.

**Comment by knutwannheden** [comment]
@seonwoo960000 Just for your understanding: The tests in rewrite-java-test don't get executed against all parser versions, so when changing the parser we try to create tests in rewrite-java-tck instead.

**Review by knutwannheden — APPROVED** [review]

**Comment by seonwoo960000** [comment]
Thank you for the clarification 😄 , helped me understand better.

**Review by traceyyoshima — APPROVED** [review]

**Pull request merged**